### ISSUE TITLE
Bug (?) - mist download installer -help doesn't display help

### ISSUE THREAD

**Issue by carlashley** [ninxsoft/mist-cli#160]
### :beetle: Description

When running `mist download firmware -help` or `mist download installer -help` it doesn't display help as `mist download -help` might.

### :clipboard: Steps to Reproduce

At the command line, either `mist download firmware -help` or `mist download installer -help` produces this output:
```
[nemo@42wallabyway]:~ # mist download firmware -help
Error: Missing value for '-e <export>'
Help:  -e <export>  Specify the path to export the download results to one of the following formats:
* /path/to/export.json (JSON file)
* /path/to/export.plist (Property List file)
* /path/to/export.yaml (YAML file)
The following variables will be dynamically substituted:
* %NAME% will be replaced with 'macOS Sonoma'
* %VERSION% will be replaced with '14.0'
* %BUILD% will be replaced with '23A344'
Note: The file extension will determine the output file format.
Note: Parent directories will be created automatically.
Usage: mist download firmware [<options>] <search-string>
  See 'mist download firmware --help' for more information.
[nemo@42wallabyway]:~ # mist download installer -help
Error: Missing value for '-e <export>'
Help:  -e <export>  Specify the path to export the download results to one of the following formats:
* /path/to/export.json (JSON file)
* /path/to/export.plist (Property List file)
* /path/to/export.yaml (YAML file)
The following variables will be dynamically substituted:
* %NAME% will be replaced with 'macOS Sonoma'
* %VERSION% will be replaced with '14.0'
* %BUILD% will be replaced with '23A344'
Note: The file extension will determine the output file format.
Note: Parent directories will be created automatically.
Usage: mist download installer [<options>] <search-string> <output-type> ...
  See 'mist download installer --help' for more information.
[nemo@42wallabyway]:~ # 
```

### :white_check_mark: Expected Behaviour

The expected outcome is a firm scalding that `--help` is the correct argument to use, but really, if it parsed `-help` like the below command does that would also be 😄 
```
[nemo@42wallabyway]:~ # mist download -help          
OVERVIEW: Download a macOS Firmware / Installer.

USAGE: mist download <subcommand>

OPTIONS:
  --version               Show the version.
  -h, --help              Show help information.

SUBCOMMANDS:
  firmware                Download a macOS Firmware.
                          * macOS Firmwares are for Apple Silicon Macs only.
  installer               Download a macOS Installer.
                          * macOS Installers for macOS Catalina 10.15 and older are for Intel based Macs only.
                          * macOS Installers for macOS Big Sur 11 and newer are Universal - for both Apple Silicon and Intel based Macs.

  See 'mist help download <subcommand>' for detailed help.
[nemo@42wallabyway]:~ # 
```

### :computer: Environment

- mist version (`mist --version`): **2.0 (latest: 2.0)**
- macOS Version (`sw_vers`): **14.3.1 (23D60)**
- Hardware (`system_profiler SPHardwareDataType`):
  - Model Identifier: **MacBookPro18,2**
  - Chip : **Apple M1 Max**

### :camera: Screenshots

![Screen_Shot_2024-02-09-072428-550156](https://github.com/ninxsoft/mist-cli/assets/9841187/9083aba1-8aec-4a61-abcf-5f1b136129b0)

### :information_source: Additional context

- Is this also a bug in [Mist.app](https://github.com/ninxsoft/Mist)? :x:

**Comment by gregneagle** [comment]
Are you sure `-help` is supported and it shouldn't actually be `-h` or `--help`?

**Comment by gregneagle** [comment]
It's very common that command-line parsing libraries _enforce_ styles like `-h/--help`: `-help` is interpreted as "-h -e -l -p`. Asking the dev of this tool to use a different command-line parsing libraries or customize the behavior of the one he's using seems unfair.

**Comment by carlashley** [comment]
I don't believe that's what I'm asking @gregneagle, and with respect, this is @ninxsoft's issue tracker, I'm more than happy to hear from him on the matter.

**Comment by grahampugh** [comment]
I think the fact that you got away with `download -help` is just luck that there isn't a `-e` option associated with the download verb, whereas there is a `-e` value associated with the firmware/installer verbs, which require a value and therefore you violated the syntax. Therefore, I don't personally see any bug here (unless you consider it a bug to ignore invalid parameters). Of course, happy for Nindi to confirm.

**Comment by ninxsoft** [comment]
> [!IMPORTANT] 
> **tl;dr:** As @grahampugh alluded to, `-help` is working as intended. For the matter of `-help` vs. `--help`, sticking to POSIX / GNU conventions makes sense to avoid confusion 👍

I can shed some light on the inner workings of `mist-cli`:

- The arguments are parsed using [swift-argument-parser](https://github.com/apple/swift-argument-parser), which is somewhat analogous to [argparse](https://docs.python.org/3/library/argparse.html) if you're coming from the Python world 🐍
- `mist` supports a combination of POSIX (short) and GNU (long) style options, or both where it makes sense:
  - [25.1.1 Program Argument Syntax Conventions](https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html)
  - `mist download firmware --help` or `mist download installer --help` for the full list
- By default `mist` will display the relevant help message for the option / flag that has been entered incorrectly. As @gregneagle pointed out, this is the behaviour you are witnessing @carlashley when entering `-help`:
  1. Parse `-h`, circle back and print help at the end if nothing errors out
  2. Parse `-e`, throw an error as the `<export>` value is missing
  3. `-l` never gets parsed
  4. `-p` never gets parsed
- Flags / options that are not supported altogether will also throw their respective error:
  - ```shell
    mist download firmware "Sonoma" -lp
    Error: Unknown option '-lp'
    Usage: mist download firmware [<options>] <search-string>
    See 'mist download firmware --help' for more information.

    mist download firmware "Sonoma" -l
    Error: Unknown option '-l'
    Usage: mist download firmware [<options>] <search-string>
    See 'mist download firmware --help' for more information.

    mist download firmware "Sonoma" -p
    Error: Unknown option '-p'
    Usage: mist download firmware [<options>] <search-string>
    See 'mist download firmware --help' for more information.
    ```
In your case @carlashley, you're seeing just the error output for `-e`, as `mist` never gets to `-h`, `-l` or `-p`.

Regarding `mist download -help` or even `mist -help`, I suspect you are witnessing slightly different behaviour due to the lack of custom logic. I have only baked in custom logic for `mist list <subcommand>` and `mist download <subcommand>`. I would assume it's defaulting to showing help if `-h` or `--help` is specified, regardless of whether unknown options are added or not. I haven't coded up any custom logic because `mist can't really do anything without the `<subcommand>` argument 🤷
